layer_factory/CMakeLists.txt

```diff
@@ -64,51 +64,93 @@ MACRO(SUBDIRLIST result curdir)
-    file(TO_NATIVE_PATH ${JSON_DEST_PATH}/VkLayer_${target}.json dst_json)
-    file(WRITE ${dst_json} ${target_json_file})
-    # Copy target json file from build/layers to final directory at compile time
-    file(TO_NATIVE_PATH ${JSON_DECORATED_DEST_PATH}/VkLayer_${target}.json decorated_dst_json)
-    add_custom_target(${target}json-copy ALL
-        COMMAND ${CMAKE_COMMAND} -E copy ${dst_json} ${decorated_dst_json}
-        VERBATIM
-        )
-    set_target_properties(${target}json-copy PROPERTIES FOLDER ${VULKANTOOLS_TARGET_FOLDER})
 
-    add_library(VkLayer_${target} SHARED ${ARGN} VkLayer_${target}.def)
-    target_link_Libraries(VkLayer_${target} ${VkLayer_utils_LIBRARY})
-    target_include_directories(VkLayer_${target} PRIVATE ${subdir})
-    endmacro()
-else()
-    macro(add_factory_layer target subdir)
-    # Create custom target for the copies so that json template is copied and edited at compile time
-    file(READ "${JSON_TEMPLATE_PATH}" json_file_template)
-    string(REPLACE "layer_factory" "${target}" target_json_file "${json_file_template}")
-    file(TO_NATIVE_PATH ${JSON_DEST_PATH}/VkLayer_${target}.json dst_json)
-    file(WRITE ${dst_json} ${target_json_file})
-    add_library(VkLayer_${target} SHARED ${ARGN})
-    target_link_Libraries(VkLayer_${target} ${VkLayer_utils_LIBRARY})
-    target_include_directories(VkLayer_${target} PRIVATE ${subdir})
-    set_target_properties(VkLayer_${target} PROPERTIES LINK_FLAGS "-Wl,-Bsymbolic,--exclude-libs,ALL")
-    endmacro()
-endif()
+
+set (CMAKE_LAYER_FACTORY_SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR})
+SUBDIRLIST(ST_SUBDIRS ${CMAKE_LAYER_FACTORY_SOURCE_DIR})
+
+# json file creation
+
+# The output file needs Unix "/" separators or Windows "\" separators
+# On top of that, Windows separators actually need to be doubled because the json format uses backslash escapes
+file(TO_NATIVE_PATH "./" RELATIVE_PATH_PREFIX)
+string(REPLACE "\\" "\\\\" RELATIVE_PATH_PREFIX "${RELATIVE_PATH_PREFIX}")
+
+# Run json.in file through the generator
+# We need to create the generator.cmake script so that the generator can be run at compile time, instead of configure time
+# Running at compile time lets us use cmake generator expressions (TARGET_FILE_NAME and TARGET_FILE_DIR, specifically)
+file(WRITE "${CMAKE_CURRENT_BINARY_DIR}/generator.cmake" "configure_file(\"\${INPUT_FILE}\" \"\${OUTPUT_FILE}\")")
+foreach(SUBDIR ${ST_SUBDIRS})
+    set(TARGET_NAME VkLayer_${SUBDIR})
+    set(CONFIG_DEFINES
+        -DINPUT_FILE=${JSON_TEMPLATE_PATH}
+        -DVK_VERSION="${VulkanHeaders_VERSION_MAJOR}.${VulkanHeaders_VERSION_MINOR}.${VulkanHeaders_VERSION_PATCH}"
+    )
+    set(CONFIG_DEFINES ${CONFIG_DEFINES}
+        -DOUTPUT_FILE="$<TARGET_FILE_DIR:${TARGET_NAME}>/${TARGET_NAME}.json"
+        -DRELATIVE_LAYER_BINARY="${RELATIVE_PATH_PREFIX}$<TARGET_FILE_NAME:${TARGET_NAME}>"
+        -DLAYER_NAME="VK_LAYER_LUNARG_${SUBDIR}"
+    )
+    if(CMAKE_CURRENT_BINARY_DIR MATCHES ".*layer_factor.*")
```

Comment:
To be exact, should this be "factory"?  I realize the current regex will match anyway.

Reply:
Changed to "factory".